node/pkg/dal/collector/collector.go

```diff
@@ -0,0 +1,198 @@
+	}
+
+	return collector, nil
+}
+
+func (c *Collector) Start(ctx context.Context) {
+	if c.Ctx != nil {
+		log.Warn().Str("Player", "DalCollector").Msg("Collector already running, skipping start")
+		return
+	}
+
+	ctxWithCancel, cancel := context.WithCancel(ctx)
+	c.CancelFunc = cancel
+	c.Ctx = ctxWithCancel
+
+	c.receive(ctxWithCancel)
+	c.trackOracleAdded(ctxWithCancel)
+}
+
+func (c *Collector) Stop() {
+	if c.CancelFunc != nil {
+		c.CancelFunc()
+		c.Ctx = nil
+	}
+}
+
+func (c *Collector) receive(ctx context.Context) {
+	for id := range c.IncomingStream {
+		go func(id int32) {
+			if err := c.receiveEach(ctx, id); err != nil {
+				log.Error().Err(err).Str("Player", "DalCollector").Msg("Error in receiveEach goroutine")
+			}
+		}(id)
+	}
+}
+
+func (c *Collector) receiveEach(ctx context.Context, configId int32) error {
+	err := db.Subscribe(ctx, keys.SubmissionDataStreamKey(configId), c.IncomingStream[configId])
+	if err != nil {
+		return err
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case data := <-c.IncomingStream[configId]:
+			go c.processIncomingData(ctx, data)
+		}
+	}
+}
+
+func (c *Collector) processIncomingData(ctx context.Context, data aggregator.SubmissionData) {
+	result, err := c.IncomingDataToOutgoingData(ctx, data)
+	if err != nil {
+		log.Error().Err(err).Str("Player", "DalCollector").Msg("failed to convert incoming data to outgoing data")
+		return
+	}
+
+	c.OutgoingStream[data.GlobalAggregate.ConfigID] <- *result
+}
+
+func (c *Collector) IncomingDataToOutgoingData(ctx context.Context, data aggregator.SubmissionData) (*dalcommon.OutgoingSubmissionData, error) {
+	c.mu.RLock()
+	whitelist := c.CachedWhitelist
+	c.mu.RUnlock()
+	orderedProof, err := orderProof(
+		ctx,
+		data.Proof.Proof,
+		data.GlobalAggregate.Value,
+		data.GlobalAggregate.Timestamp,
+		c.Symbols[data.GlobalAggregate.ConfigID],
+		whitelist)
+	if err != nil {
+		log.Error().Err(err).Str("Player", "DalCollector").Str("Symbol", c.Symbols[data.GlobalAggregate.ConfigID]).Msg("failed to order proof")
+		if errors.Is(err, errorSentinel.ErrReporterSignerNotWhitelisted) {
+			newList, getAllOraclesErr := getAllOracles(ctx, c.chainReader, c.submissionProxyContractAddr)
+			if getAllOraclesErr != nil {
+				log.Error().Err(getAllOraclesErr).Str("Player", "DalCollector").Msg("failed to refresh oracles")
+				return nil, getAllOraclesErr
+			}
+			c.mu.Lock()
+			c.CachedWhitelist = newList
+			c.mu.Unlock()
+		}
+		return nil, err
+	}
+	return &dalcommon.OutgoingSubmissionData{
+		Symbol:        c.Symbols[data.GlobalAggregate.ConfigID],
+		Value:         strconv.FormatInt(data.GlobalAggregate.Value, 10),
+		AggregateTime: strconv.FormatInt(data.GlobalAggregate.Timestamp.Unix(), 10),
+		Proof:         orderedProof,
+		FeedHash:      [32]byte(c.FeedHashes[data.GlobalAggregate.ConfigID]),
+		Decimals:      DefaultDecimals,
+	}, nil
+}
+
+func (c *Collector) trackOracleAdded(ctx context.Context) {
+	var eventTriggered chan any
```

Comment:
if I remember correctly, this type of initialization will initialize a `nil` channel which will always block sending and receiving. To initialize a channel as non-nil:
```go
eventTriggered = make(chan any)
```
i think this shouldnt block sending and receiving
https://stackoverflow.com/questions/39015602/how-does-a-non-initialized-channel-behave

Reply:
agreed